lib/AuthStateManager.ts

```diff
@@ -0,0 +1,174 @@
+import { AuthSdkError } from './errors';
+import { AuthState, UpdateAuthStateOptions } from './types';
+import { OktaAuth } from './browser';
+import { getConsole, warn } from './util';
+const PCancelable = require('p-cancelable');
+
+export const DEFAULT_AUTH_STATE = { 
+  isPending: true,
+  isAuthenticated: false,
+  idToken: null,
+  accessToken: null,
+};
+const DEFAULT_PENDING = {
+  updateAuthStatePromise: null,
+  canceledTimes: 0
+};
+const EVENT_AUTH_STATE_CHANGE = 'authStateChange';
+const MAX_PROMISE_CANCEL_TIMES = 10;
+
+// only compare first level of authState
+const isSameAuthState = (prevState: AuthState, state: AuthState) => {
+  return prevState.isPending === state.isPending 
+    && prevState.isAuthenticated === state.isAuthenticated 
+    && JSON.stringify(prevState.idToken) === JSON.stringify(state.idToken)
+    && JSON.stringify(prevState.accessToken) === JSON.stringify(state.accessToken)
+    && prevState.error === state.error;
+};
+
+export class AuthStateManager {
+  _sdk: OktaAuth;
+  _pending: { 
+    updateAuthStatePromise: typeof PCancelable;
+    canceledTimes: number; 
+  };
+  _authState: AuthState;
+
+  constructor(sdk: OktaAuth) {
+    if (!sdk.emitter) {
+      throw new AuthSdkError('Emitter should be initialized before AuthStateManager');
+    }
+
+    this._sdk = sdk;
+    this._pending = { ...DEFAULT_PENDING };
+    this._authState = { ...DEFAULT_AUTH_STATE };
+
+    // Listen on tokenManager events to start updateState process
+    // "added" event is emitted in both add and renew process, just listen on "added" event to update auth state
+    sdk.tokenManager.on('added', (key, token) => {
+      this.updateAuthState({ event: 'added', key, token });
+    });
+    sdk.tokenManager.on('removed', (key, token) => {
+      this.updateAuthState({ event: 'removed', key, token });
+    });
+  }
+
+  getAuthState(): AuthState {
+    return this._authState;
+  }
+
+  updateAuthState({ event, key, token }: UpdateAuthStateOptions = {}): void {
```

Comment:
Can we remove `UpdateAuthStateOptions`? It appears to be only used for logging, I think it would be better to do logging from the event handler. With options being passed it makes it look as if the produced auth state object would be a function of these options, which it is not.